crates/ide/src/syntax_highlighting/highlight.rs

```diff
@@ -277,12 +282,36 @@ pub(super) fn element(
         hash((name, shadow_count))
     }
 }
-fn highlight_def(db: &RootDatabase, def: Definition) -> Highlight {
+fn highlight_def(db: &RootDatabase, krate: Option<hir::Crate>, def: Definition) -> Highlight {
     match def {
-        Definition::Macro(_) => HlTag::Symbol(SymbolKind::Macro),
-        Definition::Field(_) => HlTag::Symbol(SymbolKind::Field),
+        Definition::Macro(m) => {
+            let mut h = Highlight::new(HlTag::Symbol(SymbolKind::Macro));
+
+            if m.module(db).map(hir::Module::krate) != krate {
+                h |= HlMod::Library;
+            }
+
+            h
+        }
+        Definition::Field(field) => {
+            let mut h = Highlight::new(HlTag::Symbol(SymbolKind::Field));
+
+            if Some(field.parent_def(db).module(db).krate()) != krate {
+                h |= HlMod::Library;
+            }
+
+            h
+        }
         Definition::ModuleDef(def) => match def {
-            hir::ModuleDef::Module(_) => HlTag::Symbol(SymbolKind::Module),
+            hir::ModuleDef::Module(module) => {
+                let mut h = Highlight::new(HlTag::Symbol(SymbolKind::Module));
+
+                if Some(module.krate()) != krate {
+                    h |= HlMod::Library;
+                }
```

Comment:
I think we can reduce code duplication here and move the library modifiers out of the match as `Definition` has a `module` function itself so you should be able to add the modifier after the match instead.

Otherwise lgtm though, bors d+